Added RPS command

OculusMode opened the pull request:
Closes #613 

## Description
Added Rock paper scissors command which allows players to play it with bot.

## Format
.rps <option_name>

where option_name can be any of r, p, s, rock, scissor, paper in any case.
## Did you:
<!-- These are required when contributing. -->
<!-- Replace [ ] with [x] to mark items as done. -->

- [x] Join the [**Python Discord Community**](https://discord.gg/python)?
- [x] Read all the comments in this template?
- [x] Ensure there is an issue open, or link relevant discord discussions?
- [x] Read the [contributing guidelines](https://pythondiscord.com/pages/contributing/contributing-guidelines/)?

Shivansh-007 reviewed (changes requested):
Hello, and thanks for your contribution!

There are some lint issues with your PR, can you please fix those?

I'd also like you to try to write more descriptive commit messages. For example, rather than `minor changes` you can write `Apply grammar changes to invalid move message`, etc.

---

Since the code is relatively small, we can merge `Game` and the cog class together.

OculusMode commented:
@Shivansh-007 Thanks a lot for reviewing I have made changes you requested and resolved all the flake8 errors the PR faced.
Just a small review of changes done:

- Merged Game class and cog class
- Added Doc strings in all the methods
- Capitalised static variables
- Added missing types
- Removed dead code
- Changed @commands.groups to @commands.command
- Changed variable name to appropriate meaningful name(i.e. arg => move)
- Checked linting on local machine and tested bot in my server

Thanks, please let me know if more changes are required.

OculusMode commented on bot/exts/evergreen/rps.py:
```diff
@@ -0,0 +1,98 @@
+        game = Game(ctx.channel)
+        await game.game_start(ctx.author, move)
+
+    @guild_only()
```
Great, I'll wait for reply of core devs if its fine or not and will change if required.

Shivansh-007 replied:
cc @Xithrius @ChrisLovering

Xithrius replied:
Feel free to remove this decorator.

OculusMode replied:
@Xithrius removed guild_only and checked on local, seems good!

OculusMode commented on bot/exts/evergreen/rps.py:
```diff
@@ -0,0 +1,98 @@
+"""
+Instead of putting bunch of conditions to check winner,
+We can just manage this dictionary
+"""
```
Changed

OculusMode commented on bot/exts/evergreen/rps.py:
```diff
@@ -0,0 +1,98 @@
+    ) -> None:
+        """Helper for playing RPS."""
+        game = Game(ctx.channel)
+        await game.game_start(ctx.author, move)
```
removed helper command.

OculusMode commented on bot/exts/evergreen/rps.py:
```diff
@@ -0,0 +1,98 @@
+    @commands.group(
+        invoke_without_command=True,
+        case_insensitive=True
+    )
```
changed as requested.

Objectivitix reviewed (commented):
It might be better to add some logical spacing here to facilitate reading :D
![image](https://user-images.githubusercontent.com/79152594/120327215-e4512f80-c2bf-11eb-90e3-a4ac2394e0c5.png)

OculusMode commented:
> It might be better to add some logical spacing here to facilitate reading :D
> ![image](https://user-images.githubusercontent.com/79152594/120327215-e4512f80-c2bf-11eb-90e3-a4ac2394e0c5.png)

@Objectivitix yeah that will give better readability indeed!
Great! I'll update it soon!

Shivansh-007 commented on bot/exts/evergreen/rps.py:
```diff
@@ -0,0 +1,83 @@
+
+    @guild_only()
+    @commands.command(
+        invoke_without_command=True,
```
This is redundant.

OculusMode replied:
Made required changes.

OculusMode commented:
@Objectivitix made required changes.!

Xithrius commented on bot/exts/evergreen/rps.py:
```diff
@@ -0,0 +1,82 @@
+class RPS(commands.Cog):
+    """Rock Paper Scissor. The Classic Game!"""
+
+    @staticmethod
+    def get_winner(action_one: str, action_two: str) -> int:
+        """Returns result of match from (-1, 0, 1) as (lost, tied, won)."""
+        return WINNER_DICT[action_one][action_two]
+
+    @staticmethod
+    def make_move() -> str:
+        """Returns random move for bot from CHOICES."""
+        return choice(CHOICES)
```
Given that these functions are used only once, I don't think there's any point to separating them from where they're being called from.

OculusMode replied:
I have removed them in new commit. Thanks.

Xithrius commented on bot/exts/evergreen/rps.py:
```diff
@@ -0,0 +1,82 @@
+from discord.ext import commands
+
+from bot.bot import Bot
+
+
```
One newline will do.
```suggestion

```

Xithrius commented on bot/exts/evergreen/rps.py:
```diff
@@ -0,0 +1,82 @@
+        """Returns random move for bot from CHOICES."""
+        return choice(CHOICES)
+
+    async def game_start(self, player: Member, channel: TextChannel, action: str) -> None:
+        """
+        Check action of player, draw a move and return result.
+
+        After checking if action of player is valid, make a random move.
+        And based on the move, compare moves of both player and bot and send approprite result.
+        """
+        if not action:
+            await channel.send("Please make a move.")
+            return
+
+        action = action.lower()
+
+        if action not in CHOICES and action not in SHORT_CHOICES:
+            await channel.send(f"Invalid move. Please make move from options: {' '.join(CHOICES)}")
+            return
+
+        bot_move = self.make_move()
+        player_result = self.get_winner(action[0], bot_move[0])
+
+        if player_result == 0:
+            message_string = f"{player.mention} You and Sir Lancebot played {bot_move.upper()}, It's a tie."
+            await channel.send(message_string)
+        elif player_result == 1:
+            await channel.send(f"Sir Lancebot played {bot_move.upper()}! {player.mention} Won!")
+        else:
+            await channel.send(f"Sir Lancebot played {bot_move.upper()}! {player.mention} Lost!")
```
Why not just have the contents of this function be within the command?

Xithrius commented on bot/exts/evergreen/rps.py:
```diff
@@ -0,0 +1,82 @@
+CHOICES = ['rock', 'paper', 'scissor']
+SHORT_CHOICES = ['r', 'p', 's']
+
+# Instead of putting bunch of conditions to check winner,
+# we can just manage this dictionary
```
```suggestion
# Using a dictionary instead of conditions to check for the winner.
```

Xithrius commented on bot/exts/evergreen/rps.py:
```diff
@@ -0,0 +1,63 @@
+        # value of player_result will be from (-1, 0, 1) as (lost, tied, won).
+        player_result = WINNER_DICT[move[0]][bot_move[0]]
+
+        if player_result == 0:
+            message_string = f"{player_mention} You and Sir Lancebot played {bot_move.upper()}, It's a tie."
+            await channel.send(message_string)
+        elif player_result == 1:
+            await channel.send(f"Sir Lancebot played {bot_move.upper()}! {player_mention} Won!")
+        else:
+            await channel.send(f"Sir Lancebot played {bot_move.upper()}! {player_mention} Lost!")
```
```suggestion
        if player_result == 0:
            message_string = f"{player_mention} You and Sir Lancebot played {bot_move}, it's a tie."
            await channel.send(message_string)
        elif player_result == 1:
            await channel.send(f"Sir Lancebot played {bot_move}! {player_mention} won!")
        else:
            await channel.send(f"Sir Lancebot played {bot_move}! {player_mention} lost!")
```

Xithrius reviewed (changes requested):
We should be using `await ctx.send()` instead of `await channel.send()` after setting `channel` to `ctx.channel`.

OculusMode commented:
> We should be using `await ctx.send()` instead of `await channel.send()`.

Yeah I think we can do that as this is kinda 1 time command. Great, I'll change it asap and recheck.

anand2312 reviewed (commented):
Pretty neat, I like the dict implementation for deciding the winners 👌🏻 
Just some minor comments

anand2312 commented on bot/exts/evergreen/rps.py:
```diff
@@ -0,0 +1,63 @@
+    async def rps(self, ctx: commands.Context, move: str) -> None:
+        """Play the classic game of Rock Paper Scissor with your own sir-lancebot!"""
+        channel = ctx.channel
+        if not move:
```
Turns out this got resolved the second I commented this 😔

OculusMode replied:
Haha, I heard it from future and changed it before you mention!

ToxicKidz commented on bot/exts/evergreen/rps.py:
```diff
@@ -0,0 +1,58 @@
+        player_mention = ctx.author.mention
+
+        if move not in CHOICES and move not in SHORT_CHOICES:
+            await ctx.send(f"Invalid move. Please make move from options: {', '.join(CHOICES).upper()}.")
+            return
```
This should come out as an error
```suggestion
            raise commands.BadArgument(f"Invalid move. Please make move from options: {', '.join(CHOICES).upper()}.")
```

ToxicKidz reviewed (changes requested):
There are two things that should be changed:
1. Single quotes are being used at the top of the file, however, our [style guide](https://pythondiscord.com/pages/guides/pydis-guides/contributing/style-guide/#quote-marks) says that we should prefer double quotes over single quotes.
2. Change the spelling of `scissor` to `scissors`.

ToxicKidz commented on bot/exts/evergreen/rps.py:
```diff
@@ -0,0 +1,57 @@
+
+    @commands.command(case_insensitive=True)
+    async def rps(self, ctx: commands.Context, move: str) -> None:
+        """Play the classic game of Rock Paper Scissor with your own sir-lancebot!"""
```
```suggestion
        """Play the classic game of Rock Paper Scissors with your own sir-lancebot!"""
```

ToxicKidz commented on bot/exts/evergreen/rps.py:
```diff
@@ -0,0 +1,57 @@
+
+
+class RPS(commands.Cog):
+    """Rock Paper Scissor. The Classic Game!"""
```
```suggestion
    """Rock Paper Scissors. The Classic Game!"""
```

ToxicKidz reviewed (approved):
:shipit:

kosayoda commented:
Thanks for the PR!
